library/src/main/java/com/google/android/exoplayer/extractor/ts/MpaReader.java

```diff
@@ -0,0 +1,293 @@
+        // Do nothing.
+    }
+
+    /**
+     * Continues a read from the provided {@code source} into a given {@code target}. It's assumed
+     * that the data should be written into {@code target} starting from an offset of zero.
+     *
+     * @param source The source from which to read.
+     * @param target The target into which data is to be read.
+     * @param targetLength The target length of the read.
+     * @return Whether the target length was reached.
+     */
+    private boolean continueRead(ParsableByteArray source, byte[] target, int targetLength) {
+        int bytesToRead = Math.min(source.bytesLeft(), targetLength - bytesRead);
+        source.readBytes(target, bytesRead, bytesToRead);
+        bytesRead += bytesToRead;
+        return bytesRead == targetLength;
+    }
+
+    /**
+     * Locates the next sync word, advancing the position to the byte that immediately follows it.
+     * If a sync word was not located, the position is advanced to the limit.
+     *
+     * @param pesBuffer The buffer whose position should be advanced.
+     * @return True if a sync word position was found. False otherwise.
+     */
+    private boolean skipToNextSync(ParsableByteArray pesBuffer) {
+        byte[] mpaData = pesBuffer.data;
+        int startOffset = pesBuffer.getPosition();
+        int endOffset = pesBuffer.limit();
+        for (int i = startOffset; i < endOffset - 1; i++) {
+            int syncBits = ((mpaData[i] & 0xFF) << 8 ) | (mpaData[i + 1] & 0xFF);
+            if ((syncBits & 0xFFF0) == 0xFFF0) {
+                hasCrc = (mpaData[i + 1] & 0x1) == 0;
+                pesBuffer.setPosition(i);
+                return true;
+            }
+        }
+        pesBuffer.setPosition(endOffset);
+        return false;
+    }
+
+    /**
+     * Calculates MPEG Audio frame size
+     *
+     * @param layer The MPEG layer
+     * @param LSF Low Sample rate Format (MPEG 2)
+     * @param bitrate The bitrate in bits per second
+     * @param samplesPerSec The sampling rate in hertz
+     * @param -paddingSize
+     * @return Frame size in bytes
+     */
+    private static int CalcMpaFrameSize (int layer, int LSF, int bitrate, int samplesPerSec, int paddingSize) {
+        return (int)(Math.floor(MPA_COEFFICIENTS[LSF][layer] * bitrate / samplesPerSec) + paddingSize) * MPA_SLOT_SIZE[layer];
+    }
+
+    /**
+     * Parses the sample header.
+     */
+    private void parseHeader() {
+        int headerLength = hasCrc ? HEADER_SIZE + CRC_SIZE : HEADER_SIZE;
+
+        if (!hasOutputFormat) {
+            mpaScratch.setPosition(0);
+            mpaScratch.skipBits(12);
+            int isLSF = (!mpaScratch.readBit()) ? 1 : 0;
+            int layer = mpaScratch.readBits(2) ^ 3;
+            mpaScratch.skipBits(1);
+            int audioObjectType = 32 + layer;
+            int bitRate = MPA_BITRATES[isLSF][layer][mpaScratch.readBits(4)];
+            int sampleRate = MPA_SAMPLING_RATES[3 - isLSF][mpaScratch.readBits(2)];
```

Comment:
yep, I used the same page, the problem is that to reliable detect the header i have to store a mask of the first candidate header skip the calculate frame length and check if the next header is equal to the candidate header, if not we have to move one bit forward and try again, that´s trivial in a file but I have to see how to do it with the buffer classes that we use.